### PR TITLE
[SPARK-36210][SQL] Preserve column insertion order in Dataset.withColumns

### PR DESCRIPTION
**Pull request by koertkuipers**
### What changes were proposed in this pull request?
Preserve the insertion order of columns in Dataset.withColumns

### Why are the changes needed?
It is the expected behavior. We preserve insertion order in all other places.

### Does this PR introduce _any_ user-facing change?
No. Currently Dataset.withColumns is not actually used anywhere to insert more than one column. This change is to make sure it behaves as expected when it is used for that purpose in future.

### How was this patch tested?
Added test in DatasetSuite

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @viirya

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -2040,6 +2040,23 @@ class DatasetSuite extends QueryTest
       joined,
       (1, 1), (2, 2), (3, 3))
   }
+
+  test("SPARK-36210: withColumns preserve insertion ordering") {
+    val df = Seq(1, 2, 3).toDS()
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2045]: can we add a test with a duplicate column names?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2045]: I think `withColumns`  doesn't allow duplicate column names.

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2045]: There is one test in `DataFrameSuite`.

**Review by viirya — APPROVED** [review]

**Comment by viirya** [comment]
@HyukjinKwon Any more comments? If no, I will merge this tomorrow. Thanks.

**Comment by HyukjinKwon** [comment]
nope lgtm too!

**Comment by srowen** [comment]
Seems OK to me. I'm very slightly concerned about the behavior change - column order is different - but seems worth it as a 'fix'.

**Comment by viirya** [comment]
Thanks. Merging to master/3.2/3.1

**Comment by srowen** [comment]
I'm not super against back-porting, but is more questionable given the behavior change

**Comment by viirya** [comment]
Hmm, `withColumns` is not a public API yet. It is used by a few ML classes to add columns. Although column order is different, I guess that change is less severe? I'm totally okay to revert backported patches.

**Comment by srowen** [comment]
Oh I see, OK

**Pull request closed without merging**